src/branch_and_bound.rs

```diff
@@ -0,0 +1,861 @@
+        let list: Vec<_> =
+            select_coins_bnb(target, cost_of_change, FeeRate::ZERO, FeeRate::ZERO, &weighted_utxos)
+                .unwrap()
+                .collect();
+
+        assert_eq!(list.len(), 3);
+        assert_eq!(list[0].utxo.value, Amount::from_str("10 cBTC").unwrap());
+        assert_eq!(list[1].utxo.value, Amount::from_str("6 cBTC").unwrap());
+        assert_eq!(list[2].utxo.value, Amount::from_str("2 cBTC").unwrap());
+    }
+
+    #[test]
+    fn select_coins_bnb_exhaust() {
+        // Recreate make_hard from bitcoin core test suit.
+        // Takes 327,661 iterations to find a solution.
+        let base: usize = 2;
+        let alpha = (0..17).enumerate().map(|(i, _)| base.pow(17 + i as u32));
+        let target = Amount::from_sat(alpha.clone().sum::<usize>() as u64);
+
+        let beta = (0..17).enumerate().map(|(i, _)| {
+            let a = base.pow(17 + i as u32);
+            let b = base.pow(16 - i as u32);
+            a + b
+        });
+
+        let vals: Vec<_> = zip(alpha, beta)
+            // flatten requires iterable types.
+            // use once() to make tuple iterable.
+            .flat_map(|tup| once(tup.0).chain(once(tup.1)))
+            .map(|a| Amount::from_sat(a as u64))
+            .collect();
+
+        let weighted_utxos = create_weighted_utxos_from_values(vals);
+        let list = select_coins_bnb(
+            target,
+            Amount::ONE_SAT,
+            FeeRate::ZERO,
+            FeeRate::ZERO,
+            &weighted_utxos,
+        );
+
+        assert!(list.is_none());
+    }
+
+    #[test]
+    fn select_coins_bnb_exhaust_v2() {
+        // Takes 163,819 iterations to find a solution.
+        let base: usize = 2;
+        let mut target = 0;
+        let vals = (0..15).enumerate().flat_map(|(i, _)| {
+            let a = base.pow(15 + i as u32) as u64;
+            target += a;
+            vec![a, a + 2]
+        });
+
+        let vals: Vec<_> = vals.map(Amount::from_sat).collect();
+        let weighted_utxos = create_weighted_utxos_from_values(vals);
+        let list = select_coins_bnb(
+            Amount::from_sat(target),
+            Amount::ONE_SAT,
+            FeeRate::ZERO,
+            FeeRate::ZERO,
+            &weighted_utxos,
+        );
+
+        assert!(list.is_none());
+    }
+
+    #[test]
+    fn select_coins_bnb_exhaust_with_result() {
+        // This returns a result AND hits the iteration exhaust limit.
+
+        // Takes 163,819 iterations (hits the iteration limit).
+        let base: usize = 2;
+        let mut target = 0;
+        let vals = (0..15).enumerate().flat_map(|(i, _)| {
+            let a = base.pow(15 + i as u32) as u64;
+            target += a;
+            vec![a, a + 2]
+        });
+
+        let mut vals: Vec<_> = vals.map(Amount::from_sat).collect();
+
+        // Add a value that will match the target before iteration exhaustion occurs.
+        vals.push(Amount::from_sat(target));
```

Comment:
Looks good, nice changes. 

One nit:
Regarding the `select_coins_bnb_exhaust_with_result()` test added in https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/0114f2a2f41f1eb48bc68555b4f574041c21c98d:
Sorry, I expressed myself imprecisely. I meant that you should consider a test that finds _a solution_ but not the best solution due to hitting the iteration limit first. You have a somewhat complicated setup in this test with adding the powers-of-2 and powers-of-2+2 sats UTXOs, and I was first scratching my head a bit how you would even find any solution because your `cost_of_change` is set to 1, and therefore using any of the "+2 UTXOs" would mean that you do not find the solution. Only then did I see that you added another UTXO that exactly matches the target. That means of course that BnB finds the best solution in the first iteration and then simply exhausts the iteration limit afterwards.

I guess that works, but it also means that the exact values you add in the loop basically don’t matter and the test is unnecessarily hard to review.

Reply:
What you have already covers the main concern: I was looking to see a test confirm that the selection function will return a result even if it does not manage to conclude the search. I just mean that if you’ll have a single UTXO that immediately provides the best solution, the remainder of the test could be much simpler.